File: Университет/cp_trainings/СНМ (DSU)/ITMO CF/step1/B/B.cpp
// author: Danila "akshin_" Axyonov

#include <iostream>
#include <string>
#include <vector>
#include <unordered_map>
#include <algorithm>
using namespace std;
typedef long long int lli;

#define plli                pair<lli, lli>
#define mp(_first, _second) make_pair(_first, _second)
#define pb(_elem)           push_back(_elem)




class DSU {
private:
    vector<lli> p,
                min_,
                max_,
                size;

public:
    DSU(lli n) {
        this->p.resize(n);
        generate(this->p.begin(), this->p.end(), []() { static lli i = 0LL; return i++; });
        this->min_ = p;
        this->max_ = p;
        this->size.assign(n, 1LL);
        return;
    }

    lli get(lli a) {
        if (this->p[a] != a)
            this->p[a] = this->get(this->p[a]);
        return this->p[a];
    }

    unordered_map<string, lli> min_max_size(lli a) {
        a = this->get(a);
        unordered_map<string, lli> res;
        res["min"] = this->min_[a];
        res["max"] = this->max_[a];
        res["size"] = this->size[a];
        return res;
    }

    void union_(lli a, lli b) {
        a = this->get(a);
        b = this->get(b);
        if (a == b)
            return;
        if (this->size[a] < this->size[b])
            swap(a, b);
        this->p[b] = a;
        this->min_[a] = min(this->min_[a], this->min_[b]);
        this->max_[a] = max(this->max_[a], this->max_[b]);
        this->size[a] += this->size[b];
        return;
    }
};




void solve();

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    
    solve();
    
    return 0;
}

void solve() {
    lli n,
        q;
    cin >> n >> q;
    DSU dsu(n);
    while (q--) {
        string cmd;
        cin >> cmd;
        if (cmd == "union") {
            lli u, v;
            cin >> u >> v;
            dsu.union_(--u, --v);
        } else { // cmd == "get"
            lli a;
            cin >> a;
            auto res = dsu.min_max_size(--a);
            cout << (res["min"] + 1LL) << ' ' << (res["max"] + 1LL) << ' ' << res["size"] << '\n';
        }
    }
    return;
}

File: Университет/cp_trainings/СНМ (DSU)/ITMO CF/step1/B/B.py
class DSU:
    def __init__(self, n: int) -> None:
        self.p = [i for i in range(n)]
        self.min = self.p.copy()
        self.max = self.p.copy()
        self.size = [1] * n
        return
    
    def get(self, a: int) -> int:
        if self.p[a] != a:
            self.p[a] = self.get(self.p[a])
        return self.p[a]
    
    def min_max_size(self, a: int) -> dict[str, int]:
        a = self.get(a)
        return {
            'min': self.min[a],
            'max': self.max[a],
            'size': self.size[a]
        }
    
    def union(self, a: int, b: int) -> None:
        a = self.get(a)
        b = self.get(b)
        if a == b:
            return
        if self.size[a] < self.size[b]:
            a, b = b, a
        self.p[b] = a
        self.min[a] = min(self.min[a], self.min[b])
        self.max[a] = max(self.max[a], self.max[b])
        self.size[a] += self.size[b]
        return

def main() -> None:
    n, q = map(int, input().split())
    dsu = DSU(n)
    for _ in range(q):
        inputs = input().split()
        if inputs[0] == 'union':
            u, v = int(inputs[1]) - 1, int(inputs[2]) - 1
            dsu.union(u, v)
        else: # inputs[0] == 'get'
            a = int(inputs[1]) - 1
            t = dsu.min_max_size(a)
            print(t['min'] + 1, t['max'] + 1, t['size'])
    return

if __name__ == '__main__':
    main()
